runtime/src/utils/buffer/write.rs

```diff
@@ -2,104 +2,116 @@ use crate::{Blob, Error, RwLock};
     /// Returns the current logical size of the blob including any buffered data.
     fn size(&self) -> u64 {
-        self.position + self.buffer.len() as u64
+        self.offset + self.data.len() as u64
     }
 
-    /// Appends bytes to the internal buffer, maintaining the "buffer at tip" invariant.
+    /// Truncates the buffer to the provided `len` if it falls within the current range.
+    /// Returns `false` and resets the buffer if the truncation point precedes the current range.
     ///
-    /// If the buffer capacity would be exceeded, it is flushed first. If the data
-    /// is larger than the buffer capacity, it is written directly to the blob.
+    /// # Panics
     ///
-    /// Returns an error if the write to the underlying [Blob] fails (may be due to a `flush` of data not
-    /// related to the data being written).
-    async fn write<S: Into<StableBuf>>(&mut self, buf: S) -> Result<(), Error> {
-        // If the buffer capacity will be exceeded, flush the buffer first
-        let buf = buf.into();
-        let buf_len = buf.len();
-
-        // Flush buffer if adding this data would exceed capacity
-        if self.buffer.len() + buf_len > self.capacity {
-            self.flush().await?;
-        }
-
-        // Write directly to blob if data is larger than buffer capacity
-        if buf_len > self.capacity {
-            self.blob.write_at(buf, self.position).await?;
-            self.position += buf_len as u64;
-            return Ok(());
+    /// Panics if the truncation point is greater than the current size of the blob.
+    fn truncate(&mut self, len: u64) -> bool {
+        assert!(len <= self.size());
+        if len <= self.offset {
+            self.data.clear();
+            self.offset = len;
+            return false;
         }
+        self.data.truncate((len - self.offset) as usize);
+        true
+    }
 
-        // Append to buffer (buffer is now guaranteed to have space)
-        self.buffer.extend_from_slice(buf.as_ref());
-        Ok(())
+    /// Returns true if the buffer is empty.
+    fn is_empty(&self) -> bool {
+        self.data.is_empty()
     }
 
-    /// Flushes buffered data to the underlying [Blob] and advances the position.
+    /// Takes & returns the buffered data for flushing, setting up the buffer for a new write.
+    ///
+    /// # Panics
     ///
-    /// After flushing, the buffer is reset and positioned at the new tip.
-    /// Does nothing if the buffer is empty.
+    /// Panics if the buffer is empty.
+    fn take_buf(&mut self) -> (Vec<u8>, u64) {
+        assert!(!self.is_empty());
+        let buf = std::mem::replace(&mut self.data, Vec::with_capacity(self.capacity));
+        let offset = self.offset;
+        self.offset += buf.len() as u64;
+        (buf, offset)
+    }
+
+    /// Read any requested data that overlaps with the buffer, returning the number of bytes
+    /// remaining to be read. (Any unread bytes are always those at the beginning of the range.)
```

Comment:
This was not intuitive to me but seems right on further thought. Every read_at I've seen that returns remaining always means to continue from the end.

Wonder if it makes sense to call this something else to avoid that "instinct" 🤔 . Not sure.

Reply:
yeah this is a bit unusual. Something like "unread_prefix" might be more clear but I tried to avoid multi-word var names 😄

Reply:
Maybe something like `pull` or `extract` (seems like `merge`/`extract` play well together and you certainly won't mistake that for `read` lol)

Reply:
I really like `extract`, changed.